### ISSUE TITLE
`memcpy` occasionally copies incorrect data, causing `SIGSEGV`/`SIGBUS`

### ISSUE THREAD

**Issue by tangruize** [asterinas/asterinas#1619]
### Description

While investigating previously observed abnormal behaviors [#1540,#1590,#1603,#1606], I found that (in statically linked binaries), `memcpy` occasionally copies incorrect data, which eventually leads to `SIGSEGV`/`SIGBUS` (env: Docker `asterinas/asterinas:0.8.3` - `asterinas/asterinas:0.9.4`, commit: 52bde1721e84cb9edeff1bf92e957c7e7ebe8a89 - 885950c2a4a65486524ac53e40f6b6515d7ebd3a).

Observed faulty `memcpy` in glibc location:

https://github.com/bminor/glibc/blob/ce13ab503375d6db3d97fe19c6ccf229330d3988/csu/libc-tls.c#L186

Observed `SIGSEGV`/`SIGBUS` in glibc location due to derefencing wrong addresses that copied by `memcpy`:

https://github.com/bminor/glibc/blob/ce13ab503375d6db3d97fe19c6ccf229330d3988/ctype/ctype-info.c#L31

I don't know the root cause. Just provide a reproducible test case below.

### To Reproduce

Testcase `/root/asterinas/test/apps/static/test_0x10001358_v2.c`:

<details><summary>Click to show</summary>

```c
#include <sys/syscall.h>
#include <unistd.h>
#include <stdio.h>

#define TO_STR_(x) #x
#define TO_STR(x) TO_STR_(x)

static inline void do_syscall(long syscall_num, long arg1, long arg2, long arg3) {
    __asm__ (
        "mov %0, %%rax;"     // Move syscall number to rax
        "mov %1, %%rdi;"     // Move argument 1 to rdi
        "mov %2, %%rsi;"     // Move argument 2 to rsi
        "mov %3, %%rdx;"     // Move argument 3 to rdx
        "syscall;"           // Perform the syscall
        :
        : "r" (syscall_num), "r" (arg1), "r" (arg2), "r" (arg3)  // Input operands
        : "%rax", "%rdi", "%rsi", "%rdx"  // Clobbered registers
    );
}

int my_atoi(const char *str) {
    int result = 0;

    // Convert each character to its integer value
    while (*str >= '0' && *str <= '9') {
        result = result * 10 + (*str - '0');
        str++;
    }

    return result;
}

char *to_string(char *buf, long number) {
    int i = 0;
    int len = 0;
    char temp[20];  // Temporary buffer to store the hex digits
    
    // Handle the case where number is 0
    if (number == 0) {
        temp[len++] = '0';
    }

    // Convert the number to hexadecimal
    while (number > 0 && len < 19) {
        int digit = number & 0xF;  // Get the last 4 bits (a hexadecimal digit)
        temp[len++] = (digit < 10) ? (digit + '0') : (digit - 10 + 'a');
        number >>= 4;  // Shift the number right by 4 bits (divide by 16)
    }

    // Reverse the temporary buffer into buf
    for (i = 0; i < len; i++) {
        buf[i] = temp[len - i - 1];
    }

    buf[len++] = '\n';
    buf[len] = '\0';  // Null-terminate the string
    return buf;
}

size_t my_strlen(const char *str) {
    size_t length = 0;

    // Iterate over each character in the string until the null terminator is encountered
    while (*str != '\0') {
        length++;      // Increment the length for each character
        str++;         // Move to the next character
    }

    return length;  // Return the total length of the string
}

void print_hex(long value) {
    char buf[20] = {0};
    char *hex = to_string(buf, value);
    do_syscall(SYS_write, 1, (long)hex, my_strlen(hex));
}

int main(long *argc, long *argv, void *envp) {
    if (envp != main) {
        if ((long)argc == 1) {
            do_syscall(SYS_exit, 0, 0, 0);
        }
        const char* warn = "not hacked!\n";
        do_syscall(SYS_write, 1, (long)warn, my_strlen(warn));
        print_hex((long)envp);
        print_hex((long)main);
    }

#define DST_PTR 0x10001358
    long *dst_addr = (long *)DST_PTR;
    if (dst_addr != argc) {
        const char* failed = "wrong hack: memcpy dst_addr is not " TO_STR(DST_PTR) "!\n";
        do_syscall(SYS_write, 1, (long)failed, my_strlen(failed));
        print_hex((long)argc);
    }

#define SRC_PTR 0x4c17b0
    long *src_addr = (long *)SRC_PTR;
    if (src_addr != argv) {
        const char* failed = "wrong hack: memcpy src_addr is not " TO_STR(SRC_PTR) "!\n";
        do_syscall(SYS_write, 1, (long)failed, my_strlen(failed));
        print_hex((long)argv);
    }

#define VALUE 0x4c69a0
    long value = VALUE;
    if (*dst_addr != value || *src_addr != value) {
        const char* bug = "find bug!\n";
        do_syscall(SYS_write, 1, (long)bug, my_strlen(bug));
        print_hex(*dst_addr);
        print_hex(*src_addr);
        do_syscall(SYS_exit, 1, 0, 0);
    } else {
        do_syscall(SYS_exit, 0, 0, 0);
    }
}
```

</details>

To improve reproducibility, apply a workaround for `SIGSEGV` in the init process (see #1603):

```bash
git apply <(cat <<'EOF'
diff --git a/kernel/src/sched/priority_scheduler.rs b/kernel/src/sched/priority_scheduler.rs
index 2d6be77f..661236ad 100644
--- a/kernel/src/sched/priority_scheduler.rs
+++ b/kernel/src/sched/priority_scheduler.rs
@@ -124,7 +124,8 @@ impl<T: Sync + Send + PreemptSchedInfo + FromTask<U>, U: Sync + Send + CommonSch
         {
             Some(target_cpu)
         } else {
-            None
+            //None
+            Some(target_cpu)
         }
     }

EOF
)
```

Reproduce commands:

```bash
FILE=test_0x10001358_v2

# Step 1: Compile the program
cd /root/asterinas/test/apps/static
# Compile the program with the following options:
# - Statically linked: Dynamically linked programs do not call `__libc_setup_tls`.
# - No stack protector: Avoids accessing uninitialized %fs, which would cause SIGSEGV in the hacked program.
# - Suppress warnings About mismatched argument types for `main`.
#gcc -fno-stack-protector -Wno-main -Wall -static $FILE.c -o /root/asterinas/test/apps/static/../../build/initramfs/test/static/$FILE
cat <<'EOF' > Makefile
include ../test_common.mk
EXTRA_C_FLAGS := -static -fno-stack-protector -Wno-main
EOF
make

# Step 2: Apply the hack
cd /root/asterinas/test/build/initramfs/test/static
BUG_FUNC='<__libc_setup_tls>:'  # the function calls memcpy
LUCKY_CODE='call   401060 <_init+0x60>'  # call memcpy
OFFSET=5  # offset after the memcpy call
# Locate the file offset of the memcpy call
SKIP=$(objdump -d $FILE | grep -A100 "$BUG_FUNC" | grep "$LUCKY_CODE" | head -1 | cut -d : -f1 | sed 's/ 4/0x/')
SKIP=$(($SKIP+$OFFSET))
cp $FILE $FILE.hack
# Extract the address of the `main` function in little-endian format
MAIN_ADDR=$(readelf -a $FILE | grep -w main | tr -s ' ' | cut -d ' ' -f 3 | sed 's/^00000000//' | xxd -r -p | xxd -p | fold -w2 | tac | sed 's/^/\\x/' | tr -d '\n')
# Inject the following shellcode into the binary at the computed offset:
# - 48 c7 c2 XX XX XX XX: mov $0xXXXXXXXX,$rdx (3rd argument in x86_64 calling convention)
# - ff e2: jmp *%rdx
(echo -en "\x48\xc7\xc2$MAIN_ADDR\xff\xe2") | dd of="$FILE".hack seek=$(($SKIP)) obs=1 conv=notrunc

# Step 3: Test the program without the hack in `make run`
# Run the original program to observe SIGBUS/SIGSEGV caused by dereferencing a non-canonical/invalid virtual address.
# Example: "0x5f38353331303030" causes SIGBUS due to a non-canonical memory access.
cd test/static; for i in `seq 1000000`; do ./test_0x10001358_v2 || (echo LOOP_CNT: $i; exit 1) || break; done

# Step 4: Test the program with the hack in `make run`
# Run the hacked program to see incorrect behavior after memcpy.
# Specifically, the destination data is incorrect (expected "0x4c69a0", but is "0x5f38353331303030").
cd test/static; for i in `seq 1000000`; do ./test_0x10001358_v2.hack || (echo LOOP_CNT: $i; exit 1) || break; done
```

### Results

- Step 3: triggers `SIGBUS`/`SIGSEGV` caused by dereferencing a non-canonical/invalid virtual address.

- Step 4, prints the incorrect data that `memcpy` copied:

```txt
find bug!
5f38353331303030
4c69a0
LOOP_CNT: 1606
```

### Additional context

- Older versions (before `v0.8.3`) cannot be tested due to the yanked `multiboot2` dependency.
- `memcpy` uses `vmovdqu` and `vzeroupper` instructions.

**Comment by tangruize** [comment]
The registers at the time of triggering `SIGSEGV`/`SIGBUS` are as follows

```
(gdb) info reg
rax            0x615f646165722f2e  7016437113943109422
rbx            0x1                 1
rcx            0xfffffffffffffff0  -16
rdx            0x0                 0
rsi            0x7ffffff82c38      140737487842360
rdi            0x1                 1
rbp            0x400528            0x400528
rsp            0x7ffffff82d28      0x7ffffff82d28
r8             0x3                 3
r9             0x1000000           16777216
r10            0x68                104
r11            0x200202            2097666
r12            0x1                 1
r13            0x7ffffff82e98      140737487842968
r14            0x4c5018            5001240
r15            0x7ffffff82e10      140737487842832
rip            0x45b946            0x45b946
eflags         0x210246            [ ID RF IOPL=0 IF ZF PF ]
cs             0x53                83
ss             0x4b                75
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
fs_base        0x100013c0          268440512
gs_base        0x0                 0
k_gs_base      0xffffffff88cdb000  -1999785984
cr0            0x80010033          [ PG WP NE ET MP PE ]
cr2            0x45b930            4569392
cr3            0x76b5f000          [ PDBR=486239 PCID=0 ]
cr4            0x506e8             [ OSXSAVE FSGSBASE OSXMMEXCPT OSFXSR PGE MCE PAE DE ]
cr8            0x0                 0
efer           0xd01               [ NXE LMA LME SCE ]
xmm0           vec128 {v4_float: [0x100016a0, 0x0, 0x100016a0, 0x0], v2_double: [0x100016a0, 0x100016a0], v16_int8: [0xa0, 0x16, 0x0, 0x10, 0x0, 0x0, 0x0, 0x0, 0xa0, 0x16, 0x0, 0x10, 0x0, 0x0, 0x0, 0x0], v8_int16: [0x16a0, 0x1000, 0x0, 0x0, 0x16a0, 0x1000, 0x0, 0x0], v4_int32: [0x100016a0, 0x0, 0x100016a0, 0x0], v2_int64: [0x100016a0, 0x100016a0], uint128: 0x100016a000000000100016a0}
xmm1           vec128 {v4_float: [0x65722f2e, 0x615f6461, 0x2e726464, 0x6b6162], v2_double: [0x615f646165722f2e, 0x6b61622e726464], v16_int8: [0x2e, 0x2f, 0x72, 0x65, 0x61, 0x64, 0x5f, 0x61, 0x64, 0x64, 0x72, 0x2e, 0x62, 0x61, 0x6b, 0x0], v8_int16: [0x2f2e, 0x6572, 0x6461, 0x615f, 0x6464, 0x2e72, 0x6162, 0x6b], v4_int32: [0x65722f2e, 0x615f6461, 0x2e726464, 0x6b6162], v2_int64: [0x615f646165722f2e, 0x6b61622e726464], uint128: 0x6b61622e726464615f646165722f2e}
xmm2           vec128 {v4_float: [0xc00000, 0x0, 0x800, 0x0], v2_double: [0xc00000, 0x800], v16_int8: [0x0, 0x0, 0xc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v8_int16: [0x0, 0xc0, 0x0, 0x0, 0x800, 0x0, 0x0, 0x0], v4_int32: [0xc00000, 0x0, 0x800, 0x0], v2_int64: [0xc00000, 0x800], uint128: 0x8000000000000c00000}
xmm3           vec128 {v4_float: [0x86, 0x0, 0x0, 0x0], v2_double: [0x86, 0x0], v16_int8: [0x86, 0x0 <repeats 15 times>], v8_int16: [0x86, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x86, 0x0, 0x0, 0x0], v2_int64: [0x86, 0x0], uint128: 0x86}
xmm4           vec128 {v4_float: [0x800, 0x0, 0x0, 0x0], v2_double: [0x800, 0x0], v16_int8: [0x0, 0x8, 0x0 <repeats 14 times>], v8_int16: [0x800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x800, 0x0, 0x0, 0x0], v2_int64: [0x800, 0x0], uint128: 0x800}
xmm5           vec128 {v4_float: [0x1000000, 0x0, 0x0, 0x0], v2_double: [0x1000000, 0x0], v16_int8: [0x0, 0x0, 0x0, 0x1, 0x0 <repeats 12 times>], v8_int16: [0x0, 0x100, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x1000000, 0x0, 0x0, 0x0], v2_int64: [0x1000000, 0x0], uint128: 0x1000000}
xmm6           vec128 {v4_float: [0x7512ec88, 0x3ab73c87, 0xfff8230c, 0x7fff], v2_double: [0x3ab73c877512ec88, 0x7ffffff8230c], v16_int8: [0x88, 0xec, 0x12, 0x75, 0x87, 0x3c, 0xb7, 0x3a, 0xc, 0x23, 0xf8, 0xff, 0xff, 0x7f, 0x0, 0x0], v8_int16: [0xec88, 0x7512, 0x3c87, 0x3ab7, 0x230c, 0xfff8, 0x7fff, 0x0], v4_int32: [0x7512ec88, 0x3ab73c87, 0xfff8230c, 0x7fff], v2_int64: [0x3ab73c877512ec88, 0x7ffffff8230c], uint128: 0x7ffffff8230c3ab73c877512ec88}
xmm7           vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
xmm8           vec128 {v4_float: [0x65722f2e, 0x615f6461, 0x2e726464, 0x6b6162], v2_double: [0x615f646165722f2e, 0x6b61622e726464], v16_int8: [0x2e, 0x2f, 0x72, 0x65, 0x61, 0x64, 0x5f, 0x61, 0x64, 0x64, 0x72, 0x2e, 0x62, 0x61, 0x6b, 0x0], v8_int16: [0x2f2e, 0x6572, 0x6461, 0x615f, 0x6464, 0x2e72, 0x6162, 0x6b], v4_int32: [0x65722f2e, 0x615f6461, 0x2e726464, 0x6b6162], v2_int64: [0x615f646165722f2e, 0x6b61622e726464], uint128: 0x6b61622e726464615f646165722f2e}
xmm9           vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
xmm10          vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
xmm11          vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
xmm12          vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
xmm13          vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
xmm14          vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
xmm15          vec128 {v4_float: [0x0, 0x0, 0x0, 0x0], v2_double: [0x0, 0x0], v16_int8: [0x0 <repeats 16 times>], v8_int16: [0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0], v4_int32: [0x0, 0x0, 0x0, 0x0], v2_int64: [0x0, 0x0], uint128: 0x0}
mxcsr          0x1f80              [ IM DM ZM OM UM PM ]
```

The only issue I notice is that `$rsp` is not aligned to a 16-byte boundary, which violates the calling convention and could cause unexpected behavior.

**Comment by tangruize** [comment]
In the hacking script, if setting `OFFSET=0` to observe the behavior before `memcpy`, the bug does not occurr (though `init` might still trigger `SIGSEGV`). This strongly indicates that the issue occurs in `memcpy`.

The `memcpy` assembly snippet executed is as follows (`%rdx` is `0x20` here, thus `-0x20(%rdi,%rdx,1)` equals `(%rdi)`):

```
000000000043fac0 <__memcpy_avx_unaligned_erms>:
  43fac0:       f3 0f 1e fa             endbr64
  43fac4:       48 89 f8                mov    %rdi,%rax
  43fac7:       48 83 fa 20             cmp    $0x20,%rdx
  43facb:       72 23                   jb     43faf0 <__memcpy_avx_unaligned_erms+0x30>
  43facd:       c5 fe 6f 06             vmovdqu (%rsi),%ymm0
  43fad1:       48 83 fa 40             cmp    $0x40,%rdx
  43fad5:       0f 87 a5 00 00 00       ja     43fb80 <__memcpy_avx_unaligned_erms+0xc0>
  43fadb:       c5 fe 6f 4c 16 e0       vmovdqu -0x20(%rsi,%rdx,1),%ymm1
  43fae1:       c5 fe 7f 07             vmovdqu %ymm0,(%rdi)
  43fae5:       c5 fe 7f 4c 17 e0       vmovdqu %ymm1,-0x20(%rdi,%rdx,1)
  43faeb:       c5 f8 77                vzeroupper
  43faee:       c3                      ret
```

If a context switch occurs between `43fadb` - `43fae5` without saving and restoring `ymm1`, it will result in unpredictable behaviors. From the above GDB output, `xmm1` resembles `xmm8`, we can guess that `ymm1` and `ymm8` may share each other's data in some way in another process. This happens when instructions like `paddb  %xmm8,%xmm1` are executed, which exist in `busybox`'s asm.

If we use GDB to debug the program in Linux, we can find `xmm8==0`, indicating mismanagement of `xmm` registers in Aster.

According to the [Intel 64 and IA-32 manual](https://cdrdv2-public.intel.com/782158/325462-sdm-vol-1-2abcd-3abcd-4.pdf) (Vol. 3A 2-21):

> The processor does not automatically save the context of the x87 FPU, XMM,
and MXCSR registers on a task switch. Instead, it sets the TS flag, which
causes the processor to raise an #NM exception whenever it encounters an
x87 FPU/MMX/SSE /SSE2/SSE3 instruction in the instruction stream for the
new task (with the exception of the instructions listed above).
> The fault handler for the #NM exception can then be used to clear the TS flag (with
the CLTS instruction) and save the context of the x87 FPU, XMM, and MXCSR regis-
ters. If the task never encounters an x87 FPU/MMX/SSE/SSE2/SSE3 instruction; the
x87 FPU/MMX/SSE/SSE2/ SSE3 context is never saved.

It seems saving and restoring `xmm` registers is currently unhandled. We can add saving/restoring support or simply disable SSE instructions. To disable SSE, we need to recompile `busybox`, `libc.a` and `libc.so.6` with flags like `-mno-sse`, and add flags to test cases.

Note that [static linking with glibc is discouraged](https://stackoverflow.com/questions/57476533/why-is-statically-linking-glibc-discouraged), `musl libc` or `uclibc` maybe good candidates for static linking. However, this is another issue.

**Comment by tangruize** [comment]
This bug has been fixed in #1652. A separate PR will address the signal handling FPU save/restore issue (https://github.com/asterinas/asterinas/pull/1652#discussion_r1861408152). Closing.
